mm2src/coins/eth/eth_swap_v2/eth_maker_swap_v2.rs

```diff
@@ -0,0 +1,490 @@
+            ));
+        }
+        let maker_swap_v2_contract = self
+            .swap_v2_contracts
+            .as_ref()
+            .map(|contracts| contracts.maker_swap_v2_contract)
+            .ok_or_else(|| {
+                ValidatePaymentError::InternalError("Expected swap_v2_contracts to be Some, but found None".to_string())
+            })?;
+        validate_payment_args(args.taker_secret_hash, args.maker_secret_hash, &args.amount)
+            .map_to_mm(ValidatePaymentError::InternalError)?;
+        let maker_address = public_to_address(args.maker_pub);
+        let swap_id = self.etomic_swap_id_v2(args.time_lock, args.maker_secret_hash);
+        let maker_status = self
+            .payment_status_v2(
+                maker_swap_v2_contract,
+                Token::FixedBytes(swap_id.clone()),
+                &MAKER_SWAP_V2,
+                EthPaymentType::MakerPayments,
+                MAKER_PAYMENT_STATE_INDX,
+            )
+            .await?;
+
+        let tx_from_rpc = self
+            .transaction(TransactionId::Hash(args.maker_payment_tx.tx_hash()))
+            .await?;
+        let tx_from_rpc = tx_from_rpc.as_ref().ok_or_else(|| {
+            ValidatePaymentError::TxDoesNotExist(format!(
+                "Didn't find provided tx {:?} on ETH node",
+                args.maker_payment_tx.tx_hash()
+            ))
+        })?;
+        validate_from_to_and_status(
+            tx_from_rpc,
+            maker_address,
+            maker_swap_v2_contract,
+            maker_status,
+            MakerPaymentStateV2::PaymentSent as u8,
+        )?;
+
+        let validation_args = {
+            let amount = wei_from_big_decimal(&args.amount, self.decimals)?;
+            MakerValidationArgs {
+                swap_id,
+                amount,
+                taker: self.my_addr().await,
+                taker_secret_hash: args.taker_secret_hash,
+                maker_secret_hash: args.maker_secret_hash,
+                payment_time_lock: args.time_lock,
+            }
+        };
+        match self.coin_type {
+            EthCoinType::Eth => {
+                let function = MAKER_SWAP_V2.function(ETH_MAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_eth_maker_payment_data(&decoded, &validation_args, function, tx_from_rpc.value)?;
+            },
+            EthCoinType::Erc20 { token_addr, .. } => {
+                let function = MAKER_SWAP_V2.function(ERC20_MAKER_PAYMENT)?;
+                let decoded = decode_contract_call(function, &tx_from_rpc.input.0)?;
+                validate_erc20_maker_payment_data(&decoded, &validation_args, function, token_addr)?;
+            },
+            EthCoinType::Nft { .. } => {
+                return MmError::err(ValidatePaymentError::ProtocolNotSupported(
+                    "NFT protocol is not supported for ETH and ERC20 Swaps".to_string(),
+                ));
+            },
```

Comment:
can you move the function calls to the top of the function for early return incase of an error.
```rust 
let function = match self.coin_type {
  EthCoinType::Eth => MAKER_SWAP_V2.function(ETH_MAKER_PAYMENT)?,
  EthCoinType::Erc20 => MAKER_SWAP_V2.function(ERC20_MAKER_PAYMENT)?,
  EthCoinType::Not => // error
}
```

Reply:
Creating `function` variable early would be preferable if there were higher risk of the `function()` method failing. However, in practice with tested constants and json ABI, `Contract.function()` is highly unlikely to fail.

The current code avoids unnecessary creation of the `function` variable until it is actually needed.

Reply:
Disagree 🙂 
> Creating function variable early would be preferable if there were higher risk of the function() method failing. However, in practice with tested constants and json ABI, Contract.function() is highly unlikely to fail.

Agree
> The current code avoids unnecessary creation of the function variable until it is actually needed.